### ISSUE TITLE
move out of owned vec clone `vec[0].clone()`

### ISSUE THREAD

**Issue by Luro02** [rust-lang/rust-clippy#5044]
One might have code, that looks like this
```rust
fn takes_vec(vec: Vec<String>) -> String {
    vec[0]
}
```

as you can see they try to move the value out of the `Vec`, but the compiler does not allow this (because `vec[0]` returns `&String`).

```
error[E0507]: cannot move out of index of `std::vec::Vec<std::string::String>`
 --> src/main.rs:2:5
  |
2 |     vec[0]
  |     ^^^^^^ move occurs because value has type `std::string::String`, which does not implement the `Copy` trait
```

because a reference is returned one might be tempted to simply clone it:

```rust
fn takes_vec(vec: Vec<String>) -> String {
    vec[0].clone()
}
```

It would be better to write:

```rust
fn takes_vec(vec: Vec<String>) -> String {
    vec.into_iter().nth(0).unwrap()
}
```

**Comment by vallentin** [comment]
Personally, they feel less idiomatic to write, so I'd say this lint should be categorized as `clippy::perf`.

In contrast to how the [iter_nth_zero](https://rust-lang.github.io/rust-clippy/master/index.html#iter_nth_zero) lint is about enhancing readability. That lint would even trigger on your last example.

**Comment by vallentin** [comment]
Did a quick and small benchmark for future reference.

It seems that for small vectors then indexing and cloning consistently performs worse. However, if the vectors simply contains around or more than 1000 elements, then on average the best performing solution heavily starts fluctuating between each benchmark.

```
// vec with 1 element
test tests::bench_index_clone    ... bench:         222 ns/iter (+/- 10)
test tests::bench_into_iter_next ... bench:         154 ns/iter (+/- 6)
test tests::bench_into_iter_nth  ... bench:         156 ns/iter (+/- 9)

// vec with 10 elements
test tests::bench_index_clone    ... bench:         864 ns/iter (+/- 83)
test tests::bench_into_iter_next ... bench:         812 ns/iter (+/- 33)
test tests::bench_into_iter_nth  ... bench:         819 ns/iter (+/- 33)

// vec with 1000 elements
test tests::bench_index_clone    ... bench:      79,312 ns/iter (+/- 4,126)
test tests::bench_into_iter_next ... bench:      78,483 ns/iter (+/- 3,347)
test tests::bench_into_iter_nth  ... bench:      80,071 ns/iter (+/- 2,392)

// vec with 10_000 elements
test tests::bench_index_clone    ... bench:     729,381 ns/iter (+/- 30,599)
test tests::bench_into_iter_next ... bench:     733,870 ns/iter (+/- 31,328)
test tests::bench_into_iter_nth  ... bench:     735,377 ns/iter (+/- 22,165)

// vec with 100_000 elements
test tests::bench_index_clone    ... bench:   8,346,675 ns/iter (+/- 881,326)
test tests::bench_into_iter_next ... bench:   8,094,185 ns/iter (+/- 436,823)
test tests::bench_into_iter_nth  ... bench:   8,084,940 ns/iter (+/- 407,502)
```

**Comment by vallentin** [comment]
Had an idea for an alternative, which could be to use [`swap_remove`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.swap_remove). At face value, I'd assume this would be a better optimization, since it's an O(1) operation, that simply swaps the index with the end and decrements the length.

```rust
fn takes_vec(vec: Vec<String>) -> String {
    vec.swap_remove(0)
}
```

Here's some new benchmark results, that includes `swap_remove`:

```
// vec with 100 elements
test tests::bench_index_clone       ... bench:       9,043 ns/iter (+/- 897)
test tests::bench_into_iter_next    ... bench:       7,994 ns/iter (+/- 427)
test tests::bench_into_iter_nth     ... bench:       7,991 ns/iter (+/- 637)
test tests::bench_swap_remove_first ... bench:       7,839 ns/iter (+/- 230)

// vec with 100_000 elements
test tests::bench_index_clone       ... bench:  15,103,090 ns/iter (+/- 867,716)
test tests::bench_into_iter_next    ... bench:  14,989,450 ns/iter (+/- 1,799,817)
test tests::bench_into_iter_nth     ... bench:  14,640,200 ns/iter (+/- 2,109,573)
test tests::bench_swap_remove_first ... bench:  14,315,070 ns/iter (+/- 1,137,726)
```

I also tested it with different element types, copy vs clone, cheap vs expensive cloning. The conclusion is, that this is only an optimization, when used along with more expensive cloning. Thus if the type is copy or the clone operation is cheap. Then `.clone()` wins. If cloning is expensive, then `.clone()` loses.

Here's some benchmark results using different strings lengths:

```
// vec with 10_000 elements of "" (0 characters)
test tests::bench_index_clone       ... bench:      93,906 ns/iter (+/- 20,585)
test tests::bench_into_iter_next    ... bench:      94,473 ns/iter (+/- 4,711)
test tests::bench_into_iter_nth     ... bench:      94,949 ns/iter (+/- 5,990)
test tests::bench_swap_remove_first ... bench:      93,879 ns/iter (+/- 3,614)

// vec with 10_000 elements of "Foo Bar" (7 characters)
test tests::bench_index_clone       ... bench:     754,813 ns/iter (+/- 89,205)
test tests::bench_into_iter_next    ... bench:     754,090 ns/iter (+/- 45,193)
test tests::bench_into_iter_nth     ... bench:     755,367 ns/iter (+/- 20,271)
test tests::bench_swap_remove_first ... bench:     746,312 ns/iter (+/- 17,291)

// vec with 10_000 elements of "Lorem ipsum ..." (445 characters)
test tests::bench_index_clone       ... bench:   3,885,920 ns/iter (+/- 431,461)
test tests::bench_into_iter_next    ... bench:   2,323,510 ns/iter (+/- 285,793)
test tests::bench_into_iter_nth     ... bench:   2,310,155 ns/iter (+/- 443,483)
test tests::bench_swap_remove_first ... bench:   2,304,000 ns/iter (+/- 644,917)
```

The results fluctuate the most with short strings. But when using longer strings, then the trend is towards `swap_remove` performing better.

The size of element type and its heap allocation (if it does any) seem to make the biggest difference. Excluding `swap_remove`, then if the string is small like `"Foo Bar"` then `.clone()` sometimes won, but if it was long like a full on `"Lorem ipsum ..."` then the others won.

In conclusion, I'd have the lint suggest `swap_remove`, but I'm still a bit indecisive. The results depend on the string length. Is it safe to assume that strings are generally long vs short, or that vectors of strings are big vs small?

**Comment by flip1995** [comment]
I would summarize it like this:

| Vec\<T\> \ T |     small    |    big    |
|:--------|:------------:|:---------:|
| **small**    |   clone bad  | clone bad |
| **big**      | ¯\\\_(ツ)_/¯ | clone bad |

So if the item the vector holds, clone is always bad, since it adds too much overhead (This seems to scale with the vec size though). If it is only a small item, clone does only matter, if the vector is small. Since we can't check the size of a vector at compile time, we could assume, that a vector is rather big, since vectors of 1 item are rather unusual (?).

What irritates me is, that the runtime of bigger vectors increases, since indexing should be O(1)?

**Comment by camsteffen** [comment]
This can be an enhancement of `redundant_clone`.
